### ISSUE TITLE
[BUG] Core temperature not shown on Intel CPUs supported by coretemp before SandyBridge on Linux

### ISSUE THREAD

**Issue by ChUnterberger** [aristocratos/btop#151]
Btop does not show CPU core temperatures on Intel CPUs supported by coretemp before SandyBridge because these platforms do not support the package temperature sensor file **"temp1_input"** in the hwmon folder.

This is documented in the coretemp documentation found here.
[https://www.kernel.org/doc/html/v5.11/hwmon/coretemp.html](https://www.kernel.org/doc/html/v5.11/hwmon/coretemp.html)
> "The per-package sensor is new; as of now, it is present only in the SandyBridge platform."

The details are found in the source code.
[https://elixir.bootlin.com/linux/latest/source/drivers/hwmon/coretemp.c#L460](https://elixir.bootlin.com/linux/latest/source/drivers/hwmon/coretemp.c#L460)
```c
/*
 * Find attr number for sysfs:
 * We map the attr number to core id of the CPU
 * The attr number is always core id + 2
 * The Pkgtemp will always show up as temp1_*, if available
 */
```
This means that on platforms before SandyBridge the coretemp entries actually start with core 0 temperature at **"temp2_input"** and not with package temperature at **"temp1_input"**.

Here a sample of an Intel Core i7 860 (codename Lynnfield)
[https://ark.intel.com/content/www/us/en/ark/products/41316/intel-core-i7860-processor-8m-cache-2-80-ghz.html](https://ark.intel.com/content/www/us/en/ark/products/41316/intel-core-i7860-processor-8m-cache-2-80-ghz.html)
```shell
> sensors -u
coretemp-isa-0000
Adapter: ISA adapter
Core 0:
  temp2_input: 40.000
  temp2_max: 83.000
  temp2_crit: 99.000
  temp2_crit_alarm: 0.000
Core 1:
  temp3_input: 40.000
  temp3_max: 83.000
  temp3_crit: 99.000
  temp3_crit_alarm: 0.000
Core 2:
  temp4_input: 42.000
  temp4_max: 83.000
  temp4_crit: 99.000
  temp4_crit_alarm: 0.000
Core 3:
  temp5_input: 40.000
  temp5_max: 83.000
  temp5_crit: 99.000
  temp5_crit_alarm: 0.000
```
The btop code assumes that all temperature sensors have to start with **"temp1_input"** and will skip the folders where this is not the case.

I am creating this issue before writing a fix and sending in a pull request in oder to ask for comments how the code should handle this.

A quick fix would be to have a special handling for coretemp folders if there are entries starting with **"temp2_input"** and in this case set the starting index of the for loop running thru the sensors to 2 instead of 1. This would change the existing code the least and should not change the behavior for other users but looks kind of ugly. 

Please let me know what is your opinion on the matter.

Thank you for this awesome project.

Best regards,
Christian

**To Reproduce**
* Start btop on Intel Core architecture before Sandy Bridge
* Core temperature is not shown

**Expected behavior**
* Core temperature should be shown on CPU platforms supported by coretemp before Sandy Bridge

**Info:**
 - btop++ version: `1.1.0` | btop source commit: 044927b7e5161ba40bfd4e1943e89f4fe49b5c7c
 - Binary: self compiled
 - (If compiled) Compiler and version: g++ v11.1.0
 - Architecture: x86_64
 - Platform: Linux
 - (Linux) Kernel: 5.14.16-arch1-1
 - Terminal used: Alacritty
 - Font used: DejaVu Sans Mono

**Comment by aristocratos** [comment]
@ChUnterberger 
Great find and thanks for taking it upon yourself to create a fix! 👍🏼 

> A quick fix would be to have a special handling for coretemp folders if there are entries starting with "temp2_input" and in this case set the starting index of the for loop running thru the sensors to 2 instead of 1. This would change the existing code the least and should not change the behavior for other users but looks kind of ugly.

That should be fine, considering the `Cpu::get_sensors()` function is only run once at startup we don't really need to worry about efficiency there. So won't really be any performance impact adding any special cases in that function.

Feel free to submit the PR :)

**Comment by KFilipek** [comment]
Hi all,
I've prepared fix for another bug in `Cpu::get_sensors()`, that I found today on my i9-9940X.
Could you run this script on your side?
```sh
#!/bin/bash
declare -a StringArray=(`ls /sys/class/hwmon/hwmon1/temp*_label`)
for val in ${StringArray[@]}; do
   echo $val
   cat $val
done
```

I believe that new logic also will fix your case.

**Comment by ChUnterberger** [comment]
@KFilipek 

I do not currently have access to the Lynnfield Core i7 Machine mentioned above but i have the same behavior of coretemp in Atom BayTrail E3845.
[https://ark.intel.com/content/www/us/en/ark/products/78475/intel-atom-processor-e3845-2m-cache-1-91-ghz.html](https://ark.intel.com/content/www/us/en/ark/products/78475/intel-atom-processor-e3845-2m-cache-1-91-ghz.html)
These also do not report package temperature at temp1_*.

The code above produces
```shell
~$ ./sensors.sh
/sys/class/hwmon/hwmon1/temp2_label
Core 0
/sys/class/hwmon/hwmon1/temp3_label
Core 1
/sys/class/hwmon/hwmon1/temp4_label
Core 2
/sys/class/hwmon/hwmon1/temp5_label
Core 3
```
on this machine.

So your glob would find the cores on this Intel Atom as well.

The questions i see with any change in the code is how it would effect the different currently working CPU architectures the least. Some of the code looks like it handles special edge cases of CPUs i have not seen before.
[https://github.com/aristocratos/btop/blame/main/src/linux/btop_collect.cpp#L246-L252](https://github.com/aristocratos/btop/blame/main/src/linux/btop_collect.cpp#L246-L252)

Also i have not yet checked if a different insert order would break existing core mappings for users and that therefore results have to be sorted.

We should either not change too much or follow the sensors implementation in the python version so that least bpytop and btop are consistent.
[https://github.com/giampaolo/psutil/blob/master/psutil/_pslinux.py#L1314](https://github.com/giampaolo/psutil/blob/master/psutil/_pslinux.py#L1314)

**Comment by aristocratos** [comment]
@ChUnterberger 
The current implementation is based on the implementation used in psutil, it's just missing the glob for matching name of first sensor. So the fix should simply be to ignore if `temp1_input` is missing and try for `temp2_input` in the 3 cases where `temp1_input` is looked for.

> Some of the code looks like it handles special edge cases of CPUs i have not seen before.

Not really, it's the same folder structure just with some variations of the initial path on some distros/kernel combos.

**Comment by aristocratos** [comment]
Actually, I do have a vague memory that there could have been issues with some Xeon cpus skipping some numbers in the `tempX_input` order. So would probably be best to set the for loop scanning for `tempX_input` files to have a max of "numbers of cores" * 2 instead of stopping on file missing. And just add a `continue` if the current file is missing.

**Comment by KFilipek** [comment]
> Actually, I do have a vague memory that there could have been issues with some Xeon cpus skipping some numbers in the `tempX_input` order. So would probably be best to set the for loop scanning for `tempX_input` files to have a max of "numbers of cores" * 2 instead of stopping on file missing. And just add a `continue` if the current file is missing.

You are right, my i9-9940X based on Xeon have the same issue (missing `temp9_input`).

@ChUnterberger Can you try code from PR #158?

**Comment by ChUnterberger** [comment]
@KFilipek This patch would not fix this bug because the code above it which selects the directories to iterate over (search_paths) skips the folder if it does not find temp1_input. The code would also need to be extended to also include this directories to fix this bug.

That fs:directory_iterator results are not ordered does not seem to matter since the sensors are inserted in an unordered list anyway. 

I was thinking of using the fs:directory_iterator with std::regex_match(x, std::regex("temp[0-9]*_input") but s_contains could be preferred since the rest of the code uses it.

**Comment by aristocratos** [comment]
@ChUnterberger 
I think the 3 "temp1_input" could safely just changed to search for "temp2_input" instead, can't think of any system where coretemp would _only_ be reporting "temp1_input". In the case of a VM no sensors would be reported so no need to consider it.

@KFilipek 
I do appreciate your effort with the PR, but I'm gonna rewrite the function since it also need to consider the core mapping functionality in the case of a later started sensor ordering or for special cases where some numbers are missing.

**Comment by KFilipek** [comment]
> @ChUnterberger I think the 3 "temp1_input" could safely just be changed to search for "temp2_input" instead, can't think of any system where coretemp would _only_ be reporting "temp1_input". In the case of a VM no sensors would be reported so no need to consider it.
> 
> @KFilipek I do appreciate your effort with the PR, but I'm gonna rewrite the function since it also need to consider the core mapping functionality in the case of a later started sensor ordering or for special cases where some numbers are missing.

On another computer with `Intel(R) Core(TM) i5-6600 CPU @ 3.30GHz` I got:
```sh
/sys/class/hwmon/hwmon0/temp1_input
/sys/class/hwmon/hwmon1/temp1_crit
/sys/class/hwmon/hwmon1/temp1_crit_alarm
/sys/class/hwmon/hwmon1/temp1_input
/sys/class/hwmon/hwmon1/temp1_label
/sys/class/hwmon/hwmon1/temp1_max
/sys/class/hwmon/hwmon1/temp2_crit
/sys/class/hwmon/hwmon1/temp2_crit_alarm
/sys/class/hwmon/hwmon1/temp2_input
/sys/class/hwmon/hwmon1/temp2_label
/sys/class/hwmon/hwmon1/temp2_max
/sys/class/hwmon/hwmon1/temp3_crit
/sys/class/hwmon/hwmon1/temp3_crit_alarm
/sys/class/hwmon/hwmon1/temp3_input
/sys/class/hwmon/hwmon1/temp3_label
/sys/class/hwmon/hwmon1/temp3_max
/sys/class/hwmon/hwmon1/temp4_crit
/sys/class/hwmon/hwmon1/temp4_crit_alarm
/sys/class/hwmon/hwmon1/temp4_input
/sys/class/hwmon/hwmon1/temp4_label
/sys/class/hwmon/hwmon1/temp4_max
/sys/class/hwmon/hwmon1/temp5_crit
/sys/class/hwmon/hwmon1/temp5_crit_alarm
/sys/class/hwmon/hwmon1/temp5_input
/sys/class/hwmon/hwmon1/temp5_label
/sys/class/hwmon/hwmon1/temp5_max
```
I will change logic a bit to cover all cases, it will be helpful when you start rewriting.

**Comment by ChUnterberger** [comment]
@KFilipek Thank you, with this patch the core temperatures are correctly reported on my machines. 

Just a small change which might influence the behaviour for other users.
Instead of searching for `"_label"` search for `"_input"`.

This mirrors existing behaviours and would not exclude hwmon providers which only provide a temperature and nothing else.

The existing code provides default label values for this sensors.

In this check there also seems to be a bug in the double negative.

```c
if (not (dev_filename.starts_with("temp") and !dev_filename.ends_with("_label"))) {
   continue;
} else {
   search_paths.push_back(file.path());
   break;
}
```
This is hard to read and has a bug with the double negative `not (x and ! y)`.

I think this should be rewritten as 
```c
if (dev_filename.starts_with("temp") and dev_filename.ends_with("_input")) {
   search_paths.push_back(file.path());
   break;
}
```
Also i believe the call to `fs:exists` to be unnecessary if we get the path from a directory iterator.

**Comment by KFilipek** [comment]
> @KFilipek Thank you, with this patch the core temperatures are correctly reported on my machines.
> 
> Just a small change which might influence the behaviour for other users. Instead of searching for `"_label"` search for `"_input"`.
> 
> This mirrors existing behaviours and would not exclude hwmon providers which only provide a temperature and nothing else.
> 
> The existing code provides default label values for this sensors.
> 
> In this check there also seems to be a bug in the double negative.
> 
> ```c
> if (not (dev_filename.starts_with("temp") and !dev_filename.ends_with("_label"))) {
>    continue;
> } else {
>    search_paths.push_back(file.path());
>    break;
> }
> ```
> 
> This is hard to read and has a bug with the double negative `not (x and ! y)`.
> 
> I think this should be rewritten as
> 
> ```c
> if (dev_filename.starts_with("temp") and dev_filename.ends_with("_input")) {
>    search_paths.push_back(file.path());
>    break;
> }
> ```
> 
> Also i believe the call to `fs:exists` to be unnecessary if we get the path from a directory iterator.

@ChUnterberger You are right, please check now and I would be grateful for review here: #158 
Best regards,
Krzysztof Filipek

**Comment by aristocratos** [comment]
@KFilipek 
The fix I have mind for making it work if one or more "tempX_input" is missing, can't use `directory_iterator()`. It will need to iterate from 1 to for example `Shared::coreCount * 2` (to be safe) and check if the "tempX_input" of the current iteration exists and if so ignore the number from the "inputX_label" and just add it as "tempX" (X = size of list) to get a unbroken ordering.
`directory_iterator()` doesn't guarantee any specific order, so can't add them correctly (unless you keep 2 lists and compare at the end, which would be unnecessary work).

**Comment by KFilipek** [comment]
> @KFilipek The fix I have mind for making it work if one or more "tempX_input" is missing, can't use `directory_iterator()`. It will need to iterate from 1 to for example `Shared::coreCount * 2` (to be safe) and check if the "tempX_input" of the current iteration exists and if so ignore the number from the "inputX_label" and just add it as "tempX" (X = size of list) to get a unbroken ordering. `directory_iterator()` doesn't guarantee any specific order, so can't add them correctly (unless you keep 2 lists and compare at the end, which would be unnecessary work).

As I found we have info about `core_id` in `/proc/cpuinfo` (`get_core_mapping()`), so the easiest way is to iterate from 0 to max core_id we read and skip missing. Generally: X = core_id+1.
I was thinking about utilizing data about topology we have from cpuinfo.

**Comment by aristocratos** [comment]
@KFilipek 
See function https://github.com/aristocratos/btop/blob/e2668c60f6e3a7ccb698d6216462d59703976c82/src/linux/btop_collect.cpp#L407 which handles the mapping of core -> sensor

I thought the info from `/proc/cpuinfo` and the sensor order from `/sys/` was out of sync and therefor the cause of the issue. If your confident that `/proc/cpuinfo` will clear up the ordering, by all means use the new code and test if it's working.

**Comment by aristocratos** [comment]
@ChUnterberger is your issue solved by PR #158 as of the latest commit on it (3 days ago)?
And @KFilipek is your issue solved by your PR? In that's the case and you're done with the PR I will merge it in next release.

**Comment by KFilipek** [comment]
With PR:
![image](https://user-images.githubusercontent.com/7746856/143574469-923da17e-f6bd-4355-97c4-d71923e27c3c.png)
Without:
![image](https://user-images.githubusercontent.com/7746856/143574588-773d75f9-c9d5-4186-8439-6444d439c0aa.png)
So it works for me :)

**Comment by aristocratos** [comment]
@ChUnterberger 
Fixed by linked PR and commit 9465e9b in release v1.1.2
